symbolic-base/rust-wrapper/src/prover.rs

```diff
@@ -0,0 +1,888 @@
+    result.reverse();
+
+    DensePolynomial::from_coefficients_vec(result)
+}
+
+fn com(gs: &Vec<GAffine>, p: &DensePolynomial<ScalarField>) -> GAffine {
+    let scalars: Vec<_> = p.coeffs.iter().map(|i| i.into_bigint()).collect();
+    let opt_window_size = get_opt_window_size(log2(gs.len()));
+    VariableBaseMSM::multi_scalar_mul_custom(gs, &scalars, opt_window_size, 2048, 256, false).into()
+}
+
+fn compress(pt: &GAffine) -> Vec<u8> {
+    let mut buf = Vec::new();
+    pt.serialize_compressed(&mut buf)
+        .expect("Serialization failed");
+    buf
+}
+
+fn bytes(pt: &ScalarField) -> Vec<u8> {
+    unpack_scalar(*pt)
+}
+
+fn challenge(transcript: &Vec<u8>) -> ScalarField {
+    let digest = blake2b_224(transcript);
+    ScalarField::from_le_bytes_mod_order(&digest)
+}
+
+fn zip_with<A, B, C, F>(a: &[A], b: &[B], mut f: F) -> Vec<C>
+where
+    F: FnMut(&A, &B) -> C,
+{
+    a.iter().zip(b.iter()).map(|(x, y)| f(x, y)).collect()
+}
+
+fn zip_with3<'a, A, B, C, D, F>(v1: &'a [A], v2: &'a [B], v3: &'a [C], f: F) -> Vec<D>
+where
+    F: Fn(&A, &B, &C) -> D,
+{
+    v1.iter()
+        .zip(v2)
+        .zip(v3)
+        .map(|((a, b), c)| f(a, b, c))
+        .collect()
+}
+
+fn ifilter<T, F>(v: &[T], mut pred: F) -> Vec<T>
+where
+    T: Clone,
+    F: FnMut(usize, &T) -> bool,
+{
+    v.iter()
+        .enumerate()
+        .filter(|(i, x)| pred(*i, x))
+        .map(|(_, x)| x.clone())
+        .collect()
+}
+
+fn iterate_n<T, F>(n: usize, mut f: F, x: T) -> Vec<T>
+where
+    T: Clone,
+    F: FnMut(&T) -> T,
+{
+    let mut result = Vec::with_capacity(n);
+    let mut current = x;
+
+    for _ in 0..n {
+        result.push(current.clone());
+        current = f(&current);
+    }
+
+    result
+}
+
+fn sortByList(f: &Vec<ScalarField>, t: &Vec<ScalarField>) -> Vec<ScalarField> {
+    let order: HashMap<&ScalarField, usize> = t.iter().enumerate().map(|(i, v)| (v, i)).collect();
+    let mut f_copy = f.clone();
+    f_copy.sort_by_key(|x| order.get(x).cloned().unwrap());
+    f_copy
+}
+
+fn concat_vecs<T: Clone>(a: &[T], b: &[T]) -> Vec<T> {
+    let mut result = Vec::with_capacity(a.len() + b.len());
+    result.extend_from_slice(a);
+    result.extend_from_slice(b);
+    result
+}
+
+fn cumprod(pv: &DensePolynomial<ScalarField>) -> DensePolynomial<ScalarField> {
+    let v = &pv.coeffs;
+    let mut result = Vec::with_capacity(v.len());
+
+    result[0] = v[0].clone();
+    for i in 1..v.len() {
+        result[i] = &result[i - 1] * &v[i];
+    }
+
+    DensePolynomial::from_coefficients_vec(result)
+}
+
+fn rotR(pv: &DensePolynomial<ScalarField>) -> DensePolynomial<ScalarField> {
+    let mut v = pv.coeffs.clone();
+    v.rotate_right(1);
+    DensePolynomial::from_coefficients_vec(v)
+}
+
+fn rotL(pv: &DensePolynomial<ScalarField>) -> DensePolynomial<ScalarField> {
+    let mut v = pv.coeffs.clone();
+    v.rotate_left(1);
+    DensePolynomial::from_coefficients_vec(v)
+}
+
+trait Secret {
```

Comment:
Why do you need traits like this when you're going to have only single implementation? Why not define a function instead?